File: backend/application/auth_service.py
import os
from uuid import UUID

from dotenv import load_dotenv
from jose import JWTError, jwt
import bcrypt
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from domain.user import User
from infrastructure.db.models import UserModel
from infrastructure.db.session import get_session_factory

load_dotenv()

JWT_SECRET = os.getenv("JWT_SECRET", "gethired-dev-secret-change-in-production")
JWT_ALGORITHM = "HS256"
JWT_EXPIRES_IN = "7d"


class AuthService:
    def __init__(self):
        self._get_session_factory = get_session_factory

    async def register(self, name: str, email: str, password: str) -> dict:
        async with self._get_session_factory()() as session:
            async with session.begin():
                existing = await session.execute(
                    select(UserModel).where(UserModel.email == email.lower().strip())
                )
                if existing.scalar_one_or_none():
                    raise ValueError("Email already registered")

                user = UserModel(
                    name=name.strip(),
                    email=email.lower().strip(),
                    password_hash=bcrypt.hashpw(password[:72].encode(), bcrypt.gensalt()).decode(),
                )
                session.add(user)

        token = self._sign_token(user.id, user.email)
        return {
            "token": token,
            "user": {"id": str(user.id), "name": user.name, "email": user.email},
        }

    async def login(self, email: str, password: str) -> dict:
        async with self._get_session_factory()() as session:
            result = await session.execute(
                select(UserModel).where(UserModel.email == email.lower().strip())
            )
            user = result.scalar_one_or_none()

            if not user or not bcrypt.checkpw(password[:72].encode(), user.password_hash.encode()):
                raise ValueError("Invalid credentials")

            token = self._sign_token(user.id, user.email)
            return {
                "token": token,
                "user": {"id": str(user.id), "name": user.name, "email": user.email},
            }

    async def get_user(self, user_id: UUID) -> dict | None:
        async with self._get_session_factory()() as session:
            result = await session.execute(
                select(UserModel).where(UserModel.id == user_id)
            )
            user = result.scalar_one_or_none()
            if not user:
                return None
            return {"id": str(user.id), "name": user.name, "email": user.email}

    def verify_token(self, token: str) -> dict | None:
        try:
            payload = jwt.decode(token, JWT_SECRET, algorithms=[JWT_ALGORITHM])
            user_id = payload.get("sub")
            email = payload.get("email")
            if user_id is None:
                return None
            return {"id": user_id, "email": email}
        except JWTError:
            return None

    async def ensure_user_exists(self, user_id: str, email: str | None = None) -> None:
        async with self._get_session_factory()() as session:
            result = await session.execute(
                select(UserModel).where(UserModel.id == UUID(user_id))
            )
            if result.scalar_one_or_none():
                return
            async with session.begin():
                user = UserModel(
                    id=UUID(user_id),
                    name="Local User",
                    email=email or f"{user_id}@local.dev",
                    password_hash="local-only",
                )
                session.add(user)

    def _sign_token(self, user_id: UUID, email: str) -> str:
        return jwt.encode(
            {"sub": str(user_id), "email": email},
            JWT_SECRET,
            algorithm=JWT_ALGORITHM,
        )

File: backend-node/src/services/auth-service.ts
import bcrypt from "bcrypt";
import jwt from "jsonwebtoken";
import { getPool } from "../infrastructure/db/pool.js";
import { withTransaction } from "../infrastructure/db/transaction.js";
import { getConfig } from "../config/env.js";

const JWT_ALGORITHM = "HS256";
const JWT_EXPIRES_IN = "7d";
const BCRYPT_ROUNDS = 10;

export interface AuthUser {
  id: string;
  email: string;
}

export interface AuthResult {
  token: string;
  user: {
    id: string;
    name: string;
    email: string;
  };
}

export class AuthService {
  private secret: string;

  constructor() {
    this.secret = getConfig().JWT_SECRET;
  }

  async register(name: string, email: string, password: string): Promise<AuthResult> {
    const normalizedEmail = email.toLowerCase().trim();
    const trimmedName = name.trim();
    // Truncate password to 72 bytes (bcrypt limit)
    const hashedPassword = await bcrypt.hash(password.slice(0, 72), BCRYPT_ROUNDS);

    const result = await withTransaction(async (client) => {
      // Check for existing user
      const existing = await client.query(
        "SELECT id FROM users WHERE email = $1",
        [normalizedEmail],
      );
      if (existing.rows.length > 0) {
        throw new Error("Email already registered");
      }

      // Insert new user
      const insertResult = await client.query(
        `INSERT INTO users (id, name, email, password_hash, created_at, updated_at)
         VALUES (gen_random_uuid(), $1, $2, $3, NOW(), NOW())
         RETURNING id, name, email`,
        [trimmedName, normalizedEmail, hashedPassword],
      );

      return insertResult.rows[0];
    });

    const token = this.signToken(result.id, result.email);
    return {
      token,
      user: {
        id: result.id,
        name: result.name,
        email: result.email,
      },
    };
  }

  async login(email: string, password: string): Promise<AuthResult> {
    const normalizedEmail = email.toLowerCase().trim();

    const pool = getPool();
    const result = await pool.query(
      "SELECT id, name, email, password_hash FROM users WHERE email = $1",
      [normalizedEmail],
    );

    const user = result.rows[0];
    if (!user) {
      throw new Error("Invalid credentials");
    }

    const valid = await bcrypt.compare(password.slice(0, 72), user.password_hash);
    if (!valid) {
      throw new Error("Invalid credentials");
    }

    const token = this.signToken(user.id, user.email);
    return {
      token,
      user: {
        id: user.id,
        name: user.name,
        email: user.email,
      },
    };
  }

  async getUser(userId: string): Promise<{ id: string; name: string; email: string } | null> {
    const pool = getPool();
    const result = await pool.query(
      "SELECT id, name, email FROM users WHERE id = $1",
      [userId],
    );

    const user = result.rows[0];
    if (!user) return null;

    return {
      id: user.id,
      name: user.name,
      email: user.email,
    };
  }

  verifyToken(token: string): AuthUser | null {
    try {
      const payload = jwt.verify(token, this.secret, {
        algorithms: [JWT_ALGORITHM],
      }) as { sub?: string; email?: string };

      if (!payload.sub) return null;

      return {
        id: payload.sub,
        email: payload.email ?? "",
      };
    } catch {
      return null;
    }
  }

  async ensureUserExists(userId: string, email?: string): Promise<void> {
    await withTransaction(async (client) => {
      const existing = await client.query(
        "SELECT id FROM users WHERE id = $1",
        [userId],
      );
      if (existing.rows.length > 0) return;

      await client.query(
        `INSERT INTO users (id, name, email, password_hash, created_at, updated_at)
         VALUES ($1, $2, $3, 'local-only', NOW(), NOW())`,
        [userId, "Local User", email ?? `${userId}@local.dev`],
      );
    });
  }

  private signToken(userId: string, email: string): string {
    return jwt.sign(
      { sub: userId, email },
      this.secret,
      { algorithm: JWT_ALGORITHM, expiresIn: JWT_EXPIRES_IN },
    );
  }
}
